### DISCUSSION TITLE
`:hook   (after-init . global-company-mode)` can't enable company-mode in the scratch buffer.

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [company-mode/company-mode#1248] [General]
With the following testing file:

```emacs-lisp
#!/usr/bin/env bash
:;#https://www.reddit.com/r/emacs/comments/a7ubx5/an_interactive_executable_combined_bash_and_emacs/
:;#https://github.com/alphapapa/matrix-client.el/blob/d2ac55293c96d4c95971ed8e2a3f6f354565c5ed/matrix-client-standalone.el.sh#L64-L81
:;# emacs -Q --insert <(tail -n +$(grep -n -m 1 -x ";;Bootstrap straight" "$0" | egrep -o '^[[:digit:]]+') "$0"); exit
:;#https://github.com/hlissner/doom-emacs/blob/develop/bin/doom
:;#https://lists.gnu.org/archive/html/help-gnu-emacs/2021-08/msg00101.html
:;#https://groups.google.com/g/comp.unix.shell/c/krxzfCd_8qM/m/W2Lc5W-jCAAJ
:;#https://groups.google.com/g/comp.unix.shell/c/krxzfCd_8qM/m/ryoNs9AGCQAJ
:; emacs -Q --load $(realpath -e $0) -- "$@"; exit
:;# or  
:;# exec emacs -Q --load $(realpath -e $0) -- "$@"


;;Bootstrap straight
(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))
(straight-use-package 'use-package)
(setq straight-use-package-by-default t)


;https://github.com/astoff/comint-mime/issues/1#issuecomment-922396148
(use-package company  
  :demand t
  :hook
  (after-init . global-company-mode))
```
It seems that the `company-mode` can't be enabled automatically in the scratch buffer.

**Comment by dgutov** [comment]
Works for me.

I start Emacs in a new home directory, with the above code in `~/.emacs.d/init.el`, wait for it to finish its thing.

See the slash screen. Type `q` to exit it. This brings me to the scratch buffer.

Then I simply start typing: `(def` and see the completion popup show up.

Which of these steps doesn't work for you? Do you see anything interesting in the `*Messages*` buffer when that happens?

**Reply by hongyi-zhao** [reply]
> See the slash screen. Type `q` to exit it. This brings me to the scratch buffer.

The splash screen doesn’t appear on my machine at all, but instead it takes me directly to the scratch buffer.
 
> Then I simply start typing: `(def` and see the completion popup show up.

See the following:

![image](https://user-images.githubusercontent.com/11155854/138619501-48f8b0cd-8070-466c-85ff-89ddd74bea0b.png)

> Do you see anything interesting in the `*Messages*` buffer when that happens?

![image](https://user-images.githubusercontent.com/11155854/138619545-56e1e77d-ba97-4264-9bac-7101642c3729.png)

**Reply by dgutov** [reply]
Try starting with `HOME=/tmp/test-dir emacs`, perhaps after putting the reproduction config into `/tmp/test-dir/.emacs.d/init.el`.

**Reply by hongyi-zhao** [reply]
Thank you for your idea. Confirmed your conclusion with `$ bash /home/werner/.emacs.d/debug/.emacs.d/init.el`:

```emacs-lisp
#!/usr/bin/env bash
:;# $ realpath init.el 
:;# /home/werner/.emacs.d/debug/.emacs.d/init.el
:; HOME=$(dirname $(dirname $(realpath -e $0))) proxychains-ng-socks5 /usr/local/bin/emacs -- "$@"; exit

;;Bootstrap straight
(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))
(straight-use-package 'use-package)
(setq straight-use-package-by-default t)

(use-package company  
  :demand t
  :hook
  (after-init . global-company-mode))
```

**Reply by dgutov** [reply]
Then the next step is to figure out what creates the difference.

**Reply by hongyi-zhao** [reply]
It's caused by the call method: `emacs -Q --load $(realpath -e $0) -- "$@"; exit`.

But even the working method has a problem, as shown below:

![image](https://user-images.githubusercontent.com/11155854/138712936-01e90640-7e97-495d-af44-fece21db164a.png)

The icons symbols of company can't be displayed.

**Reply by dgutov** [reply]
> It's caused by the call method: emacs -Q --load $(realpath -e $0) -- "$@"; exit.

Did that load a particular file?

> But even the working method has a problem, as shown below:

IIRC this kind of problem was caused by an out-ot-date Straight recipe. The solution was to re-fetch the reciple from MELPA.

Consider using only the basic package.el (I don't use Straight or any similar alternatives myself).

**Reply by hongyi-zhao** [reply]
> Did that load a particular file?

It only loads the script in which the command resides, i.e., the whole script I posted aove.

> IIRC this kind of problem was caused by an out-ot-date Straight recipe. The solution was to re-fetch the reciple from MELPA.
> 
> Consider using only the basic package.el (I don't use Straight or any similar alternatives myself).

Yep. After deleting all the folders under the testing `.emacs.d` directory, re-running the script can fix this problem:

![image](https://user-images.githubusercontent.com/11155854/138792517-b47ee27d-d924-4b22-9665-49a6010a1a57.png)

**Reply by hongyi-zhao** [reply]
> Did that load a particular file?

See [here](https://groups.google.com/g/comp.unix.shell/c/3vrq7pqoG-A/m/Hnk0lsKfAwAJ) for more detailed explanation:

```
Some software uses /etc/passwd instead of the HOME variable.
It's a bug.

Possibly you can inspect 'proxychains-ng-socks5' if it's an script and
fix the bug yourself.

That bug was very common in the past. Fortunately nowadays most
software has fixed it. Ivan Shmakov commented about that in
comp.emacs a pair of years ago:

Ivan Shmakov <iv...@siamics.net> on comp.emacs
Message-ID: <8736wyx2...@violet.siamics.net>

Unfortunately, some software refers to passwd(5) data directly
instead, leaving no way for the user to override the home
directory location. In particular, this was the behavior (IIRC)
of programs based on older Glib versions. And I vaguely recall
that Emacs might have been affected as well.
```
